apps/apps.c

```diff
@@ -2716,6 +2716,125 @@ BIO *bio_open_owner(const char *filename, int format, int private)
     return NULL;
 }
 
+/*
+ * Open a file for writing, owner-read-only. The file is written in
+ * a 'write-to-temp-file-then-rename' basis.
+ */
+#define RND_SIZE 16
+
+static const BIO_METHOD *BIO_f_tempfile();
+
+void make_temp_filename(const char *orig, char *filename)
+{
+    int i, size = RND_SIZE;
+    char *p;
+    char set[] = "0123456789abcdefghijklmnopqrstuvwxyz"
+                 "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    strcpy(filename, orig);
+    p = filename + strlen(orig);
+    *p++ = '.';
+
+    srand(time(NULL));
+    while (size--) {
+        i = rand() % (sizeof(set) - 1);
+        *p++ = set[i];
+    }
+
+    *p = '\0';
+
+    return;
+}
+
+BIO *bio_open_owner_tmp(const char *filename, int format, int private)
+{
+    FILE *fp = NULL;
+    BIO *b = NULL;
+    BIO *sink = NULL;
+    BIO *filter = NULL;
+    int fd = -1, bflags, mode, textmode;
+    char *tmp_filename = NULL;
+
+    if (!private || filename == NULL || strcmp(filename, "-") == 0)
+        return bio_open_default(filename, 'w', format);
+
+    /* create tmp file name */
+    tmp_filename = OPENSSL_malloc(strlen(filename) + RND_SIZE + 2);
+    if (tmp_filename == NULL)
+        goto err;
+    make_temp_filename(filename, tmp_filename);
+
+    /* create a tmp file sink BIO */
+    mode = O_WRONLY;
+#ifdef O_CREAT
+    mode |= O_CREAT;
+#endif
+#ifdef O_TRUNC
+    mode |= O_TRUNC;
+#endif
+    textmode = istext(format);
+    if (!textmode) {
+#ifdef O_BINARY
+        mode |= O_BINARY;
+#elif defined(_O_BINARY)
+        mode |= _O_BINARY;
+#endif
+    }
+
+#ifdef OPENSSL_SYS_VMS
+    /* VMS doesn't have O_BINARY, it just doesn't make sense.  But,
+     * it still needs to know that we're going binary, or fdopen()
+     * will fail with "invalid argument"...  so we tell VMS what the
+     * context is.
+     */
+    if (!textmode)
+        fd = open(tmp_filename, mode, 0600, "ctx=bin");
+    else
+#endif
+        fd = open(tmp_filename, mode, 0600);
+    if (fd < 0)
+        goto err;
+    fp = fdopen(fd, modestr('w', format));
+    if (fp == NULL)
+        goto err;
+    bflags = BIO_CLOSE;
+    if (textmode)
+        bflags |= BIO_FP_TEXT;
+    sink = BIO_new_fp(fp, bflags);
+    if (sink == NULL)
+        goto err;
+
+    /* create the filter BIO */
+    filter = BIO_new(BIO_f_tempfile());
+    if (filter == NULL)
+        goto err;
+
+    b = BIO_push(filter, sink);
+
+    /* set real filename to filter BIO for later usage */
+    BIO_ctrl(filter, BIO_C_SET_FILENAME, 1, (void *)filename);
+    BIO_ctrl(filter, BIO_C_SET_FILENAME, 2, (void *)tmp_filename);
```

Comment:
I'm frankly a little surprised that you have the temporary file name creation outside of the BIO filter instead of inside.  Why does an external party need to deal with that?

Reply:
Hmm, I just copy the original `bio_open_owner` and hack something there, no special reasons...

Do you feel it's more proper to move this piece of logic into BIO filter?